### PR TITLE
Add suport for Matrix in RecordIO-Protobuf (#63)

### PR DESCRIPTION
**Pull request by juliodelgadoaws**
Encode DenseMatrix and SparseMatrix in probobuf Record format.
Transposed matrices are not supported.

*Issue #, if available: 63

*Description of changes: Encode DenseMatrix and SparseMatrix in probobuf Record format.
Transposed matrices are not supported.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [ x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-spark/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have updated the [changelog](https://github.com/aws/sagemaker-spark/blob/master/CHANGELOG.rst) with a description of my changes (if appropriate)
- [x] I have updated any necessary [documentation](https://github.com/aws/sagemaker-spark/blob/master/README.md) (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sagemaker-spark-sdk/src/test/scala/com/amazonaws/services/sagemaker/sparksdk/protobuf/ProtobufConverterTests.scala [hunk 1]
```diff
@@ -144,16 +219,41 @@ class ProtobufConverterTests extends FlatSpec with MockitoSugar with BeforeAndAf
     assert(recordToShape(record) == sparseVector.size)
   }
 
+  private def validateSparse(record: Record, sparseMatrix: SparseMatrix) : Unit = {
+    print(sparseMatrix)
```
Comment (andremoeller) [inline sagemaker-spark-sdk/src/test/scala/com/amazonaws/services/sagemaker/sparksdk/protobuf/ProtobufConverterTests.scala]: Could you remove the print?

File: sagemaker-spark-sdk/src/test/scala/com/amazonaws/services/sagemaker/sparksdk/protobuf/ProtobufConverterTests.scala [hunk 2]
```diff
@@ -144,16 +219,41 @@ class ProtobufConverterTests extends FlatSpec with MockitoSugar with BeforeAndAf
...
+    val recordKeysList = matrixRecordToKeysList(record)
+    val recordShape = recordToMatrixShape(record)
+
+    val rowIndicesSize = recordKeysList(0).toInt
+    val colPtrsSize = recordKeysList(1).toInt
+    val newSparseMatrix = new SparseMatrix(recordShape(0).toInt, recordShape(1).toInt,
+      recordKeysList.subList(2 + rowIndicesSize, recordKeysList.size).asScala.toArray.map(_.toInt),
+      recordKeysList.subList(2, 2 + rowIndicesSize).asScala.toArray.map(_.toInt),
+      recordValuesList.asScala.toArray.map(_.toDouble))
+    print(newSparseMatrix)
```
Comment (andremoeller) [inline sagemaker-spark-sdk/src/test/scala/com/amazonaws/services/sagemaker/sparksdk/protobuf/ProtobufConverterTests.scala]: Could you remove the print?

File: sagemaker-spark-sdk/src/main/scala/com/amazonaws/services/sagemaker/sparksdk/protobuf/ProtobufConverter.scala [hunk 3]
```diff
@@ -194,4 +212,67 @@ object ProtobufConverter {
...
+        }
+
+        featuresTensorBuilder.build()
+
+      case m: SparseMatrix =>
+        for (value <- m.values) {
+          featuresTensorBuilder.addValues(value.toFloat)
+        }
+
+        // Specify wheter the Sparse matrix is transposed
```
Comment (andremoeller) [inline sagemaker-spark-sdk/src/main/scala/com/amazonaws/services/sagemaker/sparksdk/protobuf/ProtobufConverter.scala]: Can you remove this comment? (It's never transposed)

File: sagemaker-spark-sdk/src/test/scala/com/amazonaws/services/sagemaker/sparksdk/protobuf/ProtobufConverterTests.scala (hunk 1 above)
Comment (juliodelgadoaws) [reply sagemaker-spark-sdk/src/test/scala/com/amazonaws/services/sagemaker/sparksdk/protobuf/ProtobufConverterTests.scala]: Absolutely. I left this behind during some "test debugging".

File: sagemaker-spark-sdk/src/main/scala/com/amazonaws/services/sagemaker/sparksdk/protobuf/ProtobufConverter.scala (hunk 3 above)
Comment (juliodelgadoaws) [reply sagemaker-spark-sdk/src/main/scala/com/amazonaws/services/sagemaker/sparksdk/protobuf/ProtobufConverter.scala]: Absolutely.

**Comment by yifeim** [comment]
Awesome! The example in https://github.com/aws/sagemaker-spark/issues/63 worked out. However, there seems to be an inconsistency between sparse and dense matrices?

Continuing on the original example:

```python
# this is correct
sm = df.select(F.udf(lambda *args: Matrices.sparse(9, 2, [0,2,4], np.ravel(args), np.ravel(args)).toDense(),
                     MatrixUDT())
                    ('c1', 'c2')
              .alias('features')
             )
sm.show()

sm.write.format('sagemaker').save('spark.io', mode='overwrite')

with open(glob.glob('spark.io/part-*')[0], 'rb') as f:
    print(read_records(f))
[features {
  key: "values"
  value {
    float32_tensor {
      values: 0.0
      values: 1.0
      values: 2.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 3.0
      values: 4.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      shape: 9
      shape: 2
    }
  }
}
, features {
  key: "values"
  value {
    float32_tensor {
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 5.0
      values: 6.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 0.0
      values: 7.0
      values: 8.0
      shape: 9
      shape: 2
    }
  }
}
]

# this is incorrect?
sm = df.select(F.udf(lambda *args: Matrices.sparse(9, 2, [0,2,4], np.ravel(args), np.ravel(args)),
                     MatrixUDT())
                    ('c1', 'c2')
              .alias('features')
             )
sm.show()

sm.write.format('sagemaker').save('spark.io', mode='overwrite')

with open(glob.glob('spark.io/part-*')[0], 'rb') as f:
    print(read_records(f))

[features {
  key: "values"
  value {
    float32_tensor {
      values: 1.0
      values: 2.0
      values: 3.0
      values: 4.0
      keys: 4
      keys: 3
      keys: 1
      keys: 2
      keys: 3
      keys: 4
      keys: 0
      keys: 2
      keys: 4
      shape: 9
      shape: 2
    }
  }
}
, features {
  key: "values"
  value {
    float32_tensor {
      values: 5.0
      values: 6.0
      values: 7.0
      values: 8.0
      keys: 4
      keys: 3
      keys: 5
      keys: 6
      keys: 7
      keys: 8
      keys: 0
      keys: 2
      keys: 4
      shape: 9
      shape: 2
    }
  }
}
]
```

**Comment by juliodelgadoaws** [comment]
I've checked out the results and they match the encoding of Sparse matrices:

See https://github.com/aws/sagemaker-spark/pull/68/files#diff-4a25a60c7160028e6f8a4c76223cd342R226
 
The Sparse matrix has to be “encoded” into the Record. The schema is:
 
https://github.com/aws/sagemaker-spark/pull/68/files#diff-4a25a60c7160028e6f8a4c76223cd342R220
 
* SparseMatrix is stored as Compressed Sparse Column (CSC)
  (See Spark SparseMatrix documentation)
   Record.values([value_0, value_1, value_2, ...])
   Record.shape([rows, cols])
   Record.keys([rowIndices_size, colPtrs_size, rowIndicies, colPtrs])

**Comment by juliodelgadoaws** [comment]
@yifeim provided the following example. I'll change the encoding format:

Let’s take an example. Say the matrix is
[0,3,4,0,0,0,0]
[0,0,0,0,0,7,8]
 
It has shape (2,7). The sparse vector representation is:
 
```
[features {
  key: "values"
  value {
    float32_tensor {
      values: 3.0
      values: 4.0
      keys: 1
      keys: 2
      shape: 7
    }
  }
}
, features {
  key: "values"
  value {
    float32_tensor {
      values: 7.0
      values: 8.0
      keys: 5
      keys: 6
      shape: 7
    }
  }
}
]
```
If we make it a sparse matrix, the representation should be:
 
```
[features {
  key: "values"
  value {
    float32_tensor {
      values: 3.0
      values: 4.0
      values: 7.0
      values: 8.0
      keys: 1
      keys: 2
      keys: 12 # floor(12/7) gives the row 1, and 12%7 gives the column 5
      keys: 13 # floor(13/7) gives the row 1, and 12%7 gives the column 6
      shape: 2 # number of rows
      shape: 7 # number of columns
    }
  }
}
```
The format also extends to higher-dimensional objects (if supported by spark).

**Comment by juliodelgadoaws** [comment]
I'll be sending a new revision today in which SparseMatrix is encoded in a different way. This is the way originally intended in the Record-protobuf format.

**Comment by yifeim** [comment]
Thanks! I verify that I can reproduce the example using the following codes:

```
df = spark.createDataFrame([[[
    [0,3,4,0,0,0,0],
    [0,0,0,0,0,7,8],
]]], ['X'])
df.show()

sm = df.select(F.udf(lambda X: Matrices.dense(2,7,np.ravel(np.transpose(X))).toSparse(),
                MatrixUDT())('X')
          .alias('features'))
sm.show()

sm.repartition(1).write.format('sagemaker').save('spark.io', mode='overwrite')
with open(glob.glob('spark.io/part-*')[0], 'rb') as f:
    print(read_records(f))
```

**Comment by ChoiByungWook** [comment]
Errors were do to coverage being out of date. That should be resolved. Resolving conflicts and running again.

**Pull request merged**